### ISSUE TITLE
Simple Get API requests sometimes slow

### ISSUE THREAD

**Issue by iquito** [elastic/elasticsearch#17451]
**Elasticsearch version**:

2.3.0

**JVM version**:

Java 8 Update 77

**OS version**:

Debian Jessie

**Description of the problem**:

I have intermittent but reproduceable slow requests when I get a record via /index/type/id and if there is a "larger" substructure (array of objects with multiple different fields and types) in the _source result. 1700 bytes (in JSON notation) is already enough to slow down a steady number of requests to above 100ms instead of the usual 1-2ms per request.

These slow requests disappear in my reproducible tests if I `_source_exclude` the substructure in my request, but they still appear randomly on the site I am using ES on (a simple ES GET then takes 200ms instead of 1-2ms). The slow requests also disappear from my reproducible tests when I start flooding ES with these simple GET requests and do additional requests in parallel. I have had slow requests of this kind since ES 1.4 and now still in ES 2.3.

I have described my experiences in detail on http://stackoverflow.com/questions/36272601/find-causes-for-slower-elasticsearch-responses - the server is idle with plenty of resources, which is why this vexes me so much. Either this is a bug, or something very strange is going on.

As far as I can tell the number of slow requests increases the larger the substructures are in the stored object, but only if they are not excluded in the request. Maybe it also has to do something with `"index": "no"`, because all the data which causes the problems is not indexed or stored, it is only in the _source.

**Comment by clintongormley** [comment]
Hi @iquito 

Requests for any stored fields (including `_id`, `_source` etc) need to decompress at least one 16kB block to extract the desired fields.  As you say, after the first request (or second request, depending on how many shards you have) the stored fields will be in the file system cache and should be fast to access. The fact that you're seeing such variability points to either (a) issues with GC or (b) hardware issues, where something else is interfering with the request.

Either way, this doesn't sound like a bug in Elasticsearch.  The better forum to talk about issues like this is http://discuss.elastic.co/ - there you'll find people who have had to deal with similar problems.

**Comment by iquito** [comment]
Thanks, I created a discussion there: https://discuss.elastic.co/t/get-api-requests-sometimes-slow-up-to-200ms-instead-of-2ms/46301

The symptoms I am experiencing are just so over the place and only concern ES (and I switched all servers around just to try it out - so the server hardware should not be a factor) that I do not know what else to try. I can even reproduce a version of these slow responses, but I do not know what else I can test to find the root cause. Nothing in the Elasticsearch documentation has any hint on how to "debug" something like this.
